### PR TITLE
Add documentation about `for` used as higher ranked trait bounds

### LINKED ISSUE #55416: Docs for `for` keyword should mention `for<'a>`
Docs for `for` keyword should mention `for<'a>` (I guess issue titles still can't have code :/)

### PR DESCRIPTION
**Pull request by czipperz**
Resolves #55416

### CODE REVIEW

File: src/libstd/keyword_docs.rs [hunk 1]
```diff
@@ -286,11 +286,13 @@ mod fn_keyword { }
 //
 /// The `for` keyword.
 ///
-/// `for` is primarily used in for-in-loops, but it has a few other pieces of syntactic uses such as
-/// `impl Trait for Type` (see [`impl`] for more info on that). for-in-loops, or to be more
-/// precise, iterator loops, are a simple syntactic sugar over an exceedingly common practice
-/// within Rust, which is to loop over an iterator until that iterator returns `None` (or `break`
-/// is called).
+/// `for` is primarily used in for-in-loops, but it has a few other pieces of syntactic
```
Comment (Centril) [inline src/libstd/keyword_docs.rs]:
```suggestion
/// `for` is used in for-in-loops, but it is used in other syntactic locations as well.
```
I don't think the assertion of "primarily" is necessarily true actually; `impl Trait for Type` is probably quite common as well.

File: src/libstd/keyword_docs.rs [hunk 2]
```diff
@@ -286,11 +286,13 @@ mod fn_keyword { }
...
+/// uses.  `for` is used when implementing traits as in `impl Trait for Type` (see
+/// [`impl`] for more info on that).  `for` is also used for [higher-ranked trait bounds]
+/// as in `for<'a> &'a T: PartialEq<i32>`.  for-in-loops, or to be more precise, iterator
+/// loops, are a simple syntactic sugar over an exceedingly common practice within Rust,
```
Comment (Centril) [inline src/libstd/keyword_docs.rs]:
```suggestion
/// loops, are a simple syntactic sugar over a common practice within Rust,
```

**Review by Centril — COMMENTED** [review]
Hmm... I think the whole paragraph needs a redo...
Can you itemize as a list and avoid starting each sentence with "`for` is used..."?

**Comment by Centril** [comment]
Looks great; @bors r+ rollup

**Comment by czipperz** [comment]
Hold on r- this I forgot to add the link @Centril

File: src/libstd/keyword_docs.rs [hunk 3]
```diff
@@ -352,6 +352,7 @@ mod fn_keyword { }
 /// For more information on for-loops, see the [Rust book] or the [Reference].
 ///
 /// [`impl`]: keyword.impl.html
+/// [higher-ranked trait bounds]: https://doc.rust-lang.org/nightly/reference/trait-bounds.html#higher-ranked-trait-bounds
```
Comment (czipperz) [inline src/libstd/keyword_docs.rs]: Should this not have `nightly/`?

**Review by czipperz — COMMENTED** [review]
Apparently this line is too long in either case.  Thoughts?

File: src/libstd/keyword_docs.rs (hunk 3 above)
Comment (Centril) [reply src/libstd/keyword_docs.rs]: Yeah you can remove that.

Comment (czipperz) [reply src/libstd/keyword_docs.rs]: It's too long even if I remove it.  Do I just put the link on the next line?  I'm not sure how Markdown would handle that.

Comment (Centril) [reply src/libstd/keyword_docs.rs]: I think moving it down should work

Comment (czipperz) [reply src/libstd/keyword_docs.rs]: Oh yeah literally 2 lines down it does that lol

**Pull request merged**